### PR TITLE
tests: fix gcc -fanalyzer errors

### PR DESCRIPTION
**Pull request by outscale-mgo**
GCC 10 introduces -fanalyzer CFLAGS, which seems to detect some errors
in the tests.

It seems gcc think those variables can be NULL when non NULL is expected,

I don't know if those are false positive, but gcc stop complaining after those
checks.

The error appear with gcc 10.1.0, using:
./configure CFLAGS="-g -O0 -fanalyzer" LDFLAGS="-fanalyzer" --enable-debug

### CODE REVIEW

**Comment by emilengler** [comment]
I honestly don't think this change is worth it. Also I don't see any logical reason why we need an fprintf there? I can imagine gcc needs this somehow as then nearly all C programs would not pass through it

File: tests/server/sws.c [hunk 1]
```diff
@@ -2064,6 +2064,10 @@ int main(int argc, char *argv[])
   case AF_UNIX:
     memset(&me.sau, 0, sizeof(me.sau));
     me.sau.sun_family = AF_UNIX;
+    if(!unix_socket) {
+      logmsg("unix_socket is unexpectedly NULL");
+      goto sws_cleanup;
+    }
```
Comment (bagder) [inline tests/server/sws.c:2070]: Can you elaborate on how this can happen? I don't see it. (I personally still avoid gcc-10's -fanalyzer option because it hasn't been very good in my tests)

**Comment by outscale-mgo** [comment]
> imagine

More than a fprintf, I had a NULL check and an early return,

As I've say I don't know if the variables can really be NULL here, but gcc seems to think so

I'll try to investigate a little more to see if they're false positive

File: tests/server/sws.c (hunk 1 above)
Comment (outscale-mgo) [reply tests/server/sws.c:2070]: I past here the 3 gcc errors, sorry if they spam a little, I'll try to investigate a little more to see if they are false positive or not:
```
sws.c: In function ‘main’:                                                                                                                                                                                                          [292/1477]
sws.c:2067:5: warning: use of NULL ‘unix_socket’ where non-null expected [CWE-690] [-Wanalyzer-null-argument]
 2067 |     strncpy(me.sau.sun_path, unix_socket, sizeof(me.sau.sun_path) - 1);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  ‘main’: events 1-12
    |
    | 1866 |   curl_socket_t sock = CURL_SOCKET_BAD;
    |      |                 ^~~~
    |      |                 |
    |      |                 (1) ‘unix_socket’ is NULL
    |......
    | 1891 |   while(argc>arg) {
    |      |        ~         
    |      |        |
    |      |        (2) following ‘false’ branch (when ‘argc <= arg’)...
    |......
    | 2015 |   install_signal_handlers(false);
    |      |   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    |      |   |
    |      |   (3) ...to here
    |......
    | 2024 |   if(CURL_SOCKET_BAD == sock) {
    |      |     ~            
    |      |     |
    |      |     (4) following ‘false’ branch (when ‘sock != -1’)...
    |......
    | 2031 |   flag = 1;
    |      |   ~~~~~~~~       
    |      |        |
    |      |        (5) ...to here
    | 2032 |   if(0 != setsockopt(sock, SOL_SOCKET, SO_REUSEADDR,
    |      |     ~            
    |      |     |
    |      |     (6) following ‘false’ branch...
    |......
    | 2039 |   if(0 != curlx_nonblock(sock, TRUE)) {
    |      |     ~     ~~~~~~~~~~~~~~~~~~~~~~~~~~
    |      |     |     |
    |      |     |     (7) ...to here
    |      |     (8) following ‘false’ branch...
    |......
    | 2046 |   switch(socket_domain) {
    |      |   ~~~~~~         
    |      |   |
    |      |   (9) ...to here
    |      |   (10) following ‘case 1:’ branch...
    |......
    | 2064 |   case AF_UNIX:
    |      |   ~~~~           
    |      |   |
    |      |   (11) ...to here
    |......
    | 2067 |     strncpy(me.sau.sun_path, unix_socket, sizeof(me.sau.sun_path) - 1);
    |      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    |      |     |
    |      |     (12) argument 2 (‘unix_socket’) NULL where non-null expected
    |
In file included from ../../lib/curl_setup_once.h:32,
                 from ../../lib/curl_setup.h:684,
                 from server_setup.h:27,
                 from sws.c:22:
/usr/include/string.h:125:14: note: argument 2 of ‘strncpy’ must be non-null
  125 | extern char *strncpy (char *__restrict __dest,
      |              ^~~~~~~
```

```
tftpd.c: In function ‘do_tftp’:
tftpd.c:964:8: warning: use of NULL ‘mode’ where non-null expected [CWE-690] [-Wanalyzer-null-argument]
  964 |     if(strcmp(pf->f_mode, mode) == 0)
      |        ^~~~~~~~~~~~~~~~~~~~~~~~
  ‘main’: events 1-16
    |
    |  547 | int main(int argc, char **argv)
    |      |     ^~~~
    |      |     |
    |      |     (1) entry to ‘main’
    |......
    |  651 |   if(CURL_SOCKET_BAD == sock) {
    |      |     ~
    |      |     |
    |      |     (2) following ‘false’ branch (when ‘sock != -1’)...
    |......
    |  659 |   flag = 1;
    |      |   ~~~~~~~~
    |      |        |
    |      |        (3) ...to here
    |  660 |   if(0 != setsockopt(sock, SOL_SOCKET, SO_REUSEADDR,
    |      |     ~
    |      |     |
    |      |     (4) following ‘false’ branch...
    |......
    |  670 |   if(!use_ipv6) {
    |      |      ~~~~~~~~~
    |      |      |
    |      |      (5) ...to here
    |......
    |  687 |   if(0 != rc) {
    |      |     ~
    |      |     |
    |      |     (6) following ‘false’ branch (when ‘rc == 0’)...
    |......
    |  695 |   if(!port) {
    |      |     ~
    |      |     |
    |      |     (7) ...to here
    |......
    |  740 |   if(!wrotepidfile) {
    |      |     ~
    |      |     |
    |      |     (8) following ‘false’ branch...
    |......
    |  745 |   if(portfile) {
    |      |      ~~~~~~~~
    |      |      |
    |      |      (9) ...to here
    |......
    |  758 |     if(!use_ipv6)
    |      |       ~
    |      |       |
    |      |       (10) following ‘true’ branch...
    |  759 | #endif
    |  760 |       fromlen = sizeof(from.sa4);
    |      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~
    |      |               |
    |      |               (11) ...to here
    |......
    |  767 |     if(got_exit_signal)
   |      |       ~
    |      |       |
    |      |       (12) following ‘false’ branch...
    |  768 |       break;
    |  769 |     if(n < 0) {
    |      |       ~
    |      |       |
    |      |       (13) ...to here
    |      |       (14) following ‘false’ branch (when ‘n >= 0’)...
    |......
    |  775 |     set_advisor_read_lock(SERVERLOGS_LOCK);
    |      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    |      |     |
    |      |     (15) ...to here
    |......
    |  779 |     if(!use_ipv6) {
    |      |       ~
    |      |       |
    |      |       (16) following ‘true’ branch...
    |
  ‘main’: events 17-20
    |
    |  781 |       from.sa4.sin_family = AF_INET;
    |      |                           ^
    |      |                           |
    |      |                           (17) ...to here
    |  782 |       peer = socket(AF_INET, SOCK_DGRAM, 0);
    |  783 |       if(CURL_SOCKET_BAD == peer) {
    |      |         ~                  
    |      |         |
    |      |         (18) following ‘false’ branch...
    |......
    |  788 |       if(connect(peer, &from.sa, sizeof(from.sa4)) < 0) {
    |      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    |      |         ||
    |      |         |(19) ...to here
    |      |         (20) following ‘false’ branch...
    |
  ‘main’: events 21-24
    |
    |  811 |     maxtimeout = 5*TIMEOUT;
    |      |                ^
    |      |                |
    |      |                (21) ...to here
    |......
    |  815 |     if(tp->th_opcode == opcode_RRQ || tp->th_opcode == opcode_WRQ) {
    |      |       ~         
    |      |       |
    |      |       (22) following ‘true’ branch...
    |  816 |       memset(&test, 0, sizeof(test));
    |      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    |      |       |
    |      |       (23) ...to here
    |  817 |       if(do_tftp(&test, tp, n) < 0)
    |      |          ~~~~~~~~~~~~~~~~~~~~~
    |      |          |
    |      |          (24) calling ‘do_tftp’ from ‘main’
    |
    +--> ‘do_tftp’: events 25-29
           |
           |  885 | static int do_tftp(struct testcase *test, struct tftphdr *tp, ssize_t size)
           |      |            ^~~~~~~
           |      |            |
           |      |            (25) entry to ‘do_tftp’
           |......
           |  888 |   int first = 1, ecode;
           |      |       ~~~~~ 
           |      |       |
           |      |       (26) ‘mode’ is NULL
           |......
           |  899 |   if(!server) {
           |      |     ~       
           |      |     |
           |      |     (27) following ‘false’ branch (when ‘server’ is non-NULL)...
           |......
           |  907 |   fprintf(server, "opcode: %x\n", tp->th_opcode);
           |      |                                   ~~~~~~~~~~~~~
           |      |                                     |
           |      |                                     (28) ...to here
           |......
           |  920 |     if(endofit)
           |      |       ~     
           |      |       |
           |      |       (29) following ‘true’ branch (when ‘endofit != 0’)...
           |
         ‘do_tftp’: event 30
           |
           |cc1:
           | (30): ...to here
           |
         ‘do_tftp’: event 31
           |
           |  947 |   if(*cp) {
           |      |     ^
           |      |     |
           |      |     (31) following ‘false’ branch...
           |
         ‘do_tftp’: event 32
           |
           |../../lib/curlx.h:85:24:
           |   85 | #define curlx_mfprintf curl_mfprintf
../../lib/curlx.h:106:18: note: in expansion of macro ‘curlx_mfprintf’
           |  106 | # define fprintf curlx_mfprintf
           |      |                  ^~~~~~~~~~~~~~
tftpd.c:954:3: note: in expansion of macro ‘fprintf’
           |  954 |   fprintf(server, "filename: %s\n", filename);
           |      |   ^~~~~~~
           |
         ‘do_tftp’: event 33
           |
           |  956 |   for(cp = mode; cp && *cp; cp++)
           |      |   ^~~
           |      |   |
           |      |   (33) following ‘false’ branch (when ‘cp’ is NULL)...
           |
         ‘do_tftp’: event 34
           |
           |../../lib/memdebug.h:154:22:
           |  154 | #define fclose(file) curl_dbg_fclose(file,__LINE__,__FILE__)
           |      |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           |      |                      |
           |      |                      (34) ...to here
tftpd.c:961:3: note: in expansion of macro ‘fclose’
           |  961 |   fclose(server);
           |      |   ^~~~~~
           |
         ‘do_tftp’: events 35-37
           |
           |  963 |   for(pf = formata; pf->f_mode; pf++)
           |      |   ^~~
           |      |   |
           |      |   (35) following ‘true’ branch...
           |  964 |     if(strcmp(pf->f_mode, mode) == 0)
           |      |        ~~~~~~~~~~~~~~~~~~~~~~~~
           |      |        |
           |      |        (36) ...to here
           |      |        (37) argument 2 (‘mode’) NULL where non-null expected
           |
In file included from ../../lib/curl_setup_once.h:32,
                 from ../../lib/curl_setup.h:684,
                 from server_setup.h:27,
                 from tftpd.c:51:
/usr/include/string.h:137:12: note: argument 2 of ‘strcmp’ must be non-null
  137 | extern int strcmp (const char *__s1, const char *__s2)
      |            ^~~~~~
```

```
unit1607.c: In function ‘test’:                                                                                                                                                                                                      [22/1477]
unit1607.c:191:13: warning: dereference of NULL ‘dns’ [CWE-690] [-Wanalyzer-null-dereference]
  191 |       if(dns->timestamp != 0) {
      |          ~~~^~~~~~~~~~~
  ‘test’: event 1
    |
    |curlcheck.h:90:7:
    |   90 |     if(unit_setup()) {                          \
    |      |       ^
    |      |       |
    |      |       (1) following ‘false’ branch...
unit1607.c:101:1: note: in expansion of macro ‘UNITTEST_START’
    |  101 | UNITTEST_START
    |      | ^~~~~~~~~~~~~~
    |
  ‘test’: events 2-7
    |
    |  104 |   int testnum = sizeof(tests) / sizeof(struct testcase);
    |      |       ^~~~~~~
    |      |       |
    |      |       (2) ...to here
    |......
    |  109 |   for(i = 0; i < testnum; ++i) {
    |      |   ~~~  
    |      |   |
    |      |   (3) following ‘true’ branch (when ‘i < testnum’)...
    |  110 |     int j;
    |  111 |     int addressnum = sizeof(tests[i].address) / sizeof(*tests[i].address);
    |      |         ~~~~~~~~~~
    |      |         |
    |      |         (4) ...to here
    |......
    |  117 |     if(!easy)
    |      |       ~
    |      |       |
    |      |       (5) following ‘false’ branch (when ‘easy’ is non-NULL)...
    |......
    |  122 |     multi = curl_multi_init();
    |      |             ~~~~~~~~~~~~~~~~~
    |      |             |
    |      |             (6) ...to here
    |......
    |  126 |     if(!list)
    |      |       ~
    |      |       |
    |      |       (7) following ‘false’ branch (when ‘list’ is non-NULL)...
    |
  ‘test’: event 8
    |
    |../../include/curl/typecheck-gcc.h:43:26:
    |   43 |       __typeof__(option) _curl_opt = option;                            \
    |      |                          ^~~~~~~~~
    |      |                          |
    |      |                          (8) ...to here
unit1607.c:128:5: note: in expansion of macro ‘curl_easy_setopt’
    |  128 |     curl_easy_setopt(easy, CURLOPT_RESOLVE, list);
    |      |     ^~~~~~~~~~~~~~~~
    |
  ‘test’: events 9-10
    |
    |  133 |     if(!entry_id)
    |      |       ^
    |      |       |
    |      |       (9) following ‘false’ branch (when ‘entry_id’ is non-NULL)...
    |  134 |       goto error;
    |  135 |     dns = Curl_hash_pick(easy->dns.hostcache, entry_id, strlen(entry_id) + 1);
    |      |                                                         ~~~~~~~~~~~~~~~~
    |      |                                                         |
    |      |                                                         (10) ...to here
    |
  ‘test’: event 11
    |
    |  139 |     addr = dns ? dns->addr : NULL;
    |      |                            ^
    |      |                            |
    |      |                            (11) following ‘false’ branch (when ‘dns’ is NULL)...
    |
  ‘test’: event 12
    |
    |
  ‘test’: events 13-42
    |
    |
```

Comment (outscale-mgo) [reply tests/server/sws.c:2070]: ok sws, seems it can't happen, because for it to happen socket_domain need to be set as AF_UNIX, and it can't happen outside of the argc parsing loop

Comment (bagder) [reply tests/server/sws.c:2070]: Yes, that was what I thought as well. And you can't silence this false positive -fanalyze warning in any neat way if I haven't missed anything. :-/

Comment (outscale-mgo) [reply tests/server/sws.c:2070]: in unit1607 gcc fail to track add properly.

Comment (outscale-mgo) [reply tests/server/sws.c:2070]: for tftpd.c: if tp->th_stuff (whish is tp->th_block) doesn't contain any '\0', mode can be use as NULL in strcmp.
I don't know how tp->th_stuff is initialize ?

Comment (outscale-mgo) [reply tests/server/sws.c:2070]: ok so to resume: sws.c and unit1607 are false positive and should not be merge, 
tftpd.c: I really don't know, I find the code a little hard to follow, so maybe it wouldn't be bad to had the check ?
So I guess I should close this PR and make a new one only for tftpd.c, do this seems like a good solution for you ?
Or maybe someone who know a little more about tftpd.c could tell me if ```tp->th_stuff``` can contain no '\0' at line 914 ?

**Comment by bagder** [comment]
I would recommend you focus such efforts on product code instead (src/lib). It's not such a big deal if there's a possible flaw in the test code. I mean, if you find a bug and fixes it that's great, it might just now be worth to hunt down questionable analyze warnings there.
